### ISSUE TITLE
SVG export loses styles of custom class when they are the same as in the body element

### ISSUE THREAD

**Issue by RainerSchwarze** [highcharts/highcharts#7687]
(Not yet sure, but another way to look at it might be "SVG export loses styles set in body element". I did not yet look into that.)

#### Expected behaviour

When I set the point marker's fill colour to black, I expect to see the same fill colour in the exported SVG.

#### Actual behaviour

When I set the point marker's fill colour to black, the fill colour in the exported SVG is the default colour (blueish).

#### Live demo with steps to reproduce

http://jsfiddle.net/v8fg80cs/

- click on chart context menu
- select download as SVG
- open SVG
- note how the fill colour of the points is blue instead of black
(If one changes both fills to "red", the output's point fill colour is again blueish. It also applies to the font.)

#### Product version

Highcharts v5 and current v6

#### Affected browser(s)

Chrome

#### Remarks

When the styles are inlined, the default browser styles are determined by adding a dummy element to the body element. This inherits all custom body styles. When the actual styles are compared against the parent element and the browser defaults (being a child of body), they are skipped, if they match.
Would it be an option to insert an iframe and add the dummy element to the iframe's body?

**Comment by codemasterover9000** [comment]
I was just about to report this issue when I found this one. Already made a fiddle: http://jsfiddle.net/jcnwfjsx/24/ basically the same as the above but with offline exporting.

It seems to affect all export types through.

**Comment by KacperMadej** [comment]
Hi @RainerSchwarze 

Default Highcharts CSS are loaded and are more precise than your CSS, that's why the color is the default one.

Use precise CSS and it works: http://jsfiddle.net/BlackLabel/7o50gj6y/

@codemasterover9000 - the same solution seems to apply for your case too: http://jsfiddle.net/BlackLabel/qet2Lnoc/

**Comment by codemasterover9000** [comment]
Is this really a solution or more of a workaround? The css is already more specific than the global highcharts.css. Proven by the fact that the browser renders using the correct colors.

The above solution at least makes the styling a lot harder.

**Comment by KacperMadej** [comment]
@codemasterover9000 In you case the `id` of container is not exported as it is only holding the chart inside. Changing CSS slightly resolves the problem: http://jsfiddle.net/BlackLabel/jcnwfjsx/25/

**Comment by codemasterover9000** [comment]
Ok thanks got it working! So basically chart.className should exist in the css root scope (not nested in some other css element).

**Comment by RainerSchwarze** [comment]
@KacperMadej Thanks for the information. 
How about the font? In my use case the font is specified site wide in the body element. I would like to see that font in the SVG export too. (The font-family entry from the highcharts.scss file has been removed so the font gets inherited.)

In my jsfiddle, when removing font-family for the body element, the SVG export gets the Courier New font. If font-family exists for body, the SVG export does not get the Courier New font.

I would assume that this is not a case for css specificity?

**Comment by KacperMadej** [comment]
Font issue looks like a bug.

Minimal demos with steps:

A. http://jsfiddle.net/v8fg80cs/
Steps:
1. export to SVG
2. font is not applied

B. (removed `font-family` from `body` in CSS) http://jsfiddle.net/BlackLabel/v8fg80cs/1/
Steps:
1. export to SVG
2. font is applied

**Comment by RainerSchwarze** [comment]
@KacperMadej I thought about this issue today and noticed, that if I change the body's fill to `#f00` in your snippet at http://jsfiddle.net/BlackLabel/7o50gj6y/, then the SVG export again has markers with blue fill. So from my point of view this is not only font related but also fill related. (and probably a couple things more)

**Comment by KacperMadej** [comment]
@RainerSchwarze 

Thank you for reporting - yes, this looks like another problem here.

In case you will find anything more please reporting it to us.

@cvasseng ?

**Comment by RainerSchwarze** [comment]
I stepped through the code which inlines the styles and watched for the fill value. The following DOM subtree shows what happens (list of classes for elements not complete):

```
default fill='#000'

g.highcharts-series-group fill='#000' (parent='#000')
  g.highcharts-series.highcharts-series-0 fill='#123' (parent='#000')
    path.highcharts-graph fill='none' (parent='#123')
  g.highcharts-marker.highcharts-series-0 fill='#123' (parent='#000')
    path.highcharts-point.highcharts-color-0 fill='#000' (parent='#123')
```

With the last path element above, the inlining code finds that the fill differs from the parent, but it is the same as the default fill and therefore not set. Which makes me wonder: The path fill differs from the g/parent fill, so it should be stored. It should not be relevant what the default fill is at this point, should it? Would the default fill test be appropriate for the subtree root only?
(Not looking at the colour issue here, just trying to understand.)

**Comment by TorsteinHonsi** [comment]
Yes, you found the sore spot there. If we just change the fill color slightly, so that it is not the same as the default, it works like expected: http://jsfiddle.net/highcharts/v8fg80cs/2/

The purpose of this check is to avoid inserting the numerous CSS properties that are default to an element, but of course if the element is different to its parent/inherited property this goes wrong. I'm just not sure how to express that in logic, as the parent may be a different type of element with different properties.

**Comment by TorsteinHonsi** [comment]
The above commit makes the font work in the exported SVG, though for some reason it doesn't seem to be respected in the PNG: http://jsfiddle.net/highcharts/v8fg80cs/3/. Most likely the "Courier" font is not available on our export server.

Offline exporting exporting however does this correctly in the PNG too, since it uses local system fonts: http://jsfiddle.net/highcharts/v8fg80cs/4/

**Comment by RainerSchwarze** [comment]
@TorsteinHonsi Is there also an issue with inheritable vs. non-inheritable style properties lying around? For instance: 

`font-weight` is inheritable, so if it differs between a node and its parent, it should be inlined. 

`border-color` (I mention this, because I used this when thinking about it) is not inheritable. 
So if it differs, it can be: 
a) specified and different: inline, 
b) different, but same as default: not inline. 
And then there's c) It is the same as with the parent element, but since it is not inheritable, it needs to be inlined. (Cases like `border-color: inherit;` should be covered then too, shouldn't it?)

Would it be necessary to know for each style property, whether it is inheritable or not?

**Comment by RainerSchwarze** [comment]
@TorsteinHonsi Another thing which might need some attention: `for (p in styles)` with `CSSStyleDeclaration` loops over array elements too. So `filterStyles(styles[p], p);` loops over p = 0, ..., 284 and then over "animation-delay", ..., "zoom". Would it make sense to only call filterStyles if p is not numeric?

**Comment by TorsteinHonsi** [comment]
Is the `border-color` property used in SVG?

> Would it make sense to only call filterStyles if p is not numeric?

Yes, that makes sense, but I just tested it now and strangely enough it doesn't lead to a performance improvement.

**Comment by RainerSchwarze** [comment]
Regarding `border-color`: I don't know. I used it to wrap my head around style inlining when playing with HTML only. I would assume that SVG has style properties which are inherited and also properties which are not inherited.

The numeric p caught my attention, when I disabled the default style check. Then the inlined styles where full of entries like "0:animation-delay;1:...". These get skipped by the default style check and then don't show up.
